lib/media/media_source_engine.js

```diff
@@ -1333,6 +1366,173 @@ shaka.media.MediaSourceEngine = class {
+   * @return {Promise<boolean>}
+   * @private
+   */
+  async determineCodecSwitch_(contentType, stream, codecSwitchingStrategy) {
+    const Capabilities = shaka.media.Capabilities;
+    const CodecSwitchingStrategy = shaka.config.CodecSwitchingStrategy;
+    const currentCodec = shaka.util.MimeUtils
+        .getCodecBase(shaka.util.MimeUtils.getCodecs(
+            this.sourceBufferTypes_[contentType]));
+    const newCodec = shaka.util.MimeUtils.getCodecBase(stream.codecs);
+
+    // Codec Switching DISABLED or current/new codecs base match
+    // then no need to switch
+    if (codecSwitchingStrategy === CodecSwitchingStrategy.DISABLED ||
+      currentCodec === newCodec) {
+      return false;
+    }
+
+    // Switching Required.
+    // Enforce Stratgies with fallback for RELOAD if SMOOTH not supported.
+    if (codecSwitchingStrategy === CodecSwitchingStrategy.SMOOTH) {
+      shaka.log.debug(`Attempting ${CodecSwitchingStrategy.SMOOTH} Strategy`);
+      if (Capabilities.isChangeTypeSupported()) {
+        await this.changeType(contentType,
+            shaka.util.MimeUtils.getFullType(stream.mimeType, stream.codecs));
+        return true;
+      } else {
+        shaka.log.debug(`${CodecSwitchingStrategy.SMOOTH}` +
+        `Strategy not supported, fallback to ${CodecSwitchingStrategy.RELOAD}`);
+      }
+    }
+    // RELOAD strategy
+    shaka.log.debug(`Using ${CodecSwitchingStrategy.RELOAD} Strategy`);
+    await this.reset(contentType, stream);
+    return true;
+  }
+
+  /**
+   * Resets the MediaSource and re-adds source buffers due to codec mismatch
+   * @param {shaka.util.ManifestParserUtils.ContentType} mismatchingType
+   * @param {shaka.extern.Stream} mismatchingStream
+   * @param {number} startTime
+   * @private
+   */
+  async reset_(mismatchingType, mismatchingStream, startTime) {
+    const ContentType = shaka.util.ManifestParserUtils.ContentType;
+
+    this.reloadingMediaSource_ = true;
+
+    try {
+      this.eventManager_.removeAll();
+      this.sourceBuffers_[ContentType.VIDEO] = null;
+      this.sourceBuffers_[ContentType.AUDIO] = null;
+
+      const previousDuration = this.mediaSource_.duration;
+      this.mediaSourceOpen_ = new shaka.util.PublicPromise();
+      this.mediaSource_ = this.createMediaSource(this.mediaSourceOpen_);
+      await this.mediaSourceOpen_;
+      this.mediaSource_.duration = previousDuration;
+
+      this.sourceBufferTypes_[mismatchingType] =
+        shaka.util.MimeUtils.getFullType(mismatchingStream.mimeType,
+            mismatchingStream.codecs);
+
+
+      const sourceBufferAdded = new shaka.util.PublicPromise();
+      const sourceBuffers =
+        /** @type {EventTarget} */(this.mediaSource_.sourceBuffers);
+
+      this.eventManager_.listenOnce(sourceBuffers, 'addsourcebuffer', () => {
+        sourceBufferAdded.resolve();
+      });
+
+      for (const contentType of [ContentType.AUDIO, ContentType.VIDEO]) {
```

Comment:
I don't think this loop is correct.  Some content may be audio-only or video-only.  You should not assume all content is audio+video.